core/blockchain.go

```diff
@@ -1807,106 +1850,121 @@ func (bc *BlockChain) insertChain(chain types.Blocks, setHead bool) (int, error)
-		snapshotAccountReadTimer.Update(statedb.SnapshotAccountReads)   // Account reads are complete(in processing)
-		snapshotStorageReadTimer.Update(statedb.SnapshotStorageReads)   // Storage reads are complete(in processing)
-		accountUpdateTimer.Update(statedb.AccountUpdates)               // Account updates are complete(in validation)
-		storageUpdateTimer.Update(statedb.StorageUpdates)               // Storage updates are complete(in validation)
-		accountHashTimer.Update(statedb.AccountHashes)                  // Account hashes are complete(in validation)
-		storageHashTimer.Update(statedb.StorageHashes)                  // Storage hashes are complete(in validation)
-		triehash := statedb.AccountHashes + statedb.StorageHashes       // The time spent on tries hashing
-		trieUpdate := statedb.AccountUpdates + statedb.StorageUpdates   // The time spent on tries update
-		trieRead := statedb.SnapshotAccountReads + statedb.AccountReads // The time spent on account read
-		trieRead += statedb.SnapshotStorageReads + statedb.StorageReads // The time spent on storage read
-		blockExecutionTimer.Update(ptime - trieRead)                    // The time spent on EVM processing
-		blockValidationTimer.Update(vtime - (triehash + trieUpdate))    // The time spent on block validation
-
-		// Write the block to the chain and get the status.
-		var (
-			wstart = time.Now()
-			status WriteStatus
-		)
-		if !setHead {
-			// Don't set the head, only insert the block
-			err = bc.writeBlockWithState(block, receipts, statedb)
-		} else {
-			status, err = bc.writeBlockAndSetHead(block, receipts, logs, statedb, false)
-		}
-		followupInterrupt.Store(true)
-		if err != nil {
-			return it.index, err
-		}
-		// Update the metrics touched during block commit
-		accountCommitTimer.Update(statedb.AccountCommits)   // Account commits are complete, we can mark them
-		storageCommitTimer.Update(statedb.StorageCommits)   // Storage commits are complete, we can mark them
-		snapshotCommitTimer.Update(statedb.SnapshotCommits) // Snapshot commits are complete, we can mark them
-		triedbCommitTimer.Update(statedb.TrieDBCommits)     // Trie database commits are complete, we can mark them
 
-		blockWriteTimer.Update(time.Since(wstart) - statedb.AccountCommits - statedb.StorageCommits - statedb.SnapshotCommits - statedb.TrieDBCommits)
-		blockInsertTimer.UpdateSince(start)
-
-		// Report the import stats before returning the various results
-		stats.processed++
-		stats.usedGas += usedGas
-
-		var snapDiffItems, snapBufItems common.StorageSize
-		if bc.snaps != nil {
-			snapDiffItems, snapBufItems = bc.snaps.Size()
-		}
-		trieDiffNodes, trieBufNodes, _ := bc.triedb.Size()
-		stats.report(chain, it.index, snapDiffItems, snapBufItems, trieDiffNodes, trieBufNodes, setHead)
+		// The traced section of block import.
+		err, stop := func() (blockEndErr error, _ bool) {
+			if bc.logger != nil {
```

Comment:
Is this whole thing just so you can properly `defer` the `bc.logger.OnBlockEnd` ? 
I think this is too large to comfortably fit in an inlined method, like this. Two suggestions

1. Move this into a proper standalone method, e.g. `err, stop := bc.processBlock(block, statedb, followupInterrupt)`
2. Try to make this method smaller. 

I think this method can be finished around here:
```golang
	blockExecutionTimer.Update(ptime - trieRead)                    // The time spent on EVM processing
	blockValidationTimer.Update(vtime - (triehash + trieUpdate))    // The time spent on block validation

	/// End this method here?

	// Write the block to the chain and get the status.
	var (
		wstart = time.Now()
		status WriteStatus
	)

```

So the actual write would then not be performed by the inlined function, but happen in the outer scope, like the code on master rigth now. The metrics and timers makes things a tad more complicated and difficult to do cleanly. 

If it doesn't turn out to work with those suggestions, then I think that a better way forward to revert this. Yes you lose the ability to defer `logger.OnBlockEnd`, so instead you'll have to invoke it at a couple of different locations, but that may be preferrable.

Reply:
I moved block processing it to a separate function. But only left updating the stats in the main func. I find it more readable now. The cut-off point was rather arbitrary.